utils/utils_test.go

```diff
@@ -45,16 +51,16 @@ func TestConvertProtocol(t *testing.T) {
 	}
 
 	for i, testCase := range testCases {
-		actualUrl, err := ConvertProtocol(testCase.urlString, testCase.protocol)
-		if err != nil && !testCase.expectedError {
-			t.Errorf("test case %d failed with unexpected error", i)
-		}
-		if err == nil && testCase.expectedError {
-			t.Errorf("test case %d did not produce expected error", i)
-		}
-		if actualUrl != testCase.expectedUrl {
-			t.Errorf("test case %d had unexpected URL. Actual: %s, Expected: %s", i, actualUrl, testCase.expectedUrl)
-		}
+		t.Run(fmt.Sprintf("test_%d", i), func(t *testing.T) {
+			actualUrl, err := ConvertProtocol(testCase.urlString, testCase.protocol)
+
+			if testCase.expectedError {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+			require.Equal(t, testCase.expectedUrl, actualUrl)
```

Comment:
do we want to move this line into the else block? since we only care about the return value when there is no error.